### ISSUE TITLE
Conv2D channels last weight layout

### ISSUE THREAD

**Issue by nsmithtt** [tenstorrent/tt-metal#13219]
## Issue
Currently the conv2d API expects channels last activations, this is totally fine and expected, but then it is mixed with channels first weights layout (i.e. PyTorch style weights).

## Proposal 
I propose we use channels last layout weights to be consistent with channels last activations, like Tensorflow, see [tf docs](https://www.tensorflow.org/api_docs/python/tf/keras/ops/conv):
> kernel has shape (kernel_spatial_shape, num_input_channels, num_output_channels)

This has the added benefit of simplifying weight preparation before moving to device (related issue):
- #12793

In many cases, if the channel dims are already tile aligned then they can just be trivially tilized.  Otherwise they just need to be padded to tile, but no permute is needed.

## Impact

Tests and models will all need to be updated to pre-prepare or create the weights in channels last layout.  Arguably, the API could support a bool denoting channels first vs last style weights for compatibility with existing tests.

**Comment by nsmithtt** [comment]
@mywoodstock, let us know your thoughts.  I think we can probably help make this change if you don't object.

**Comment by nsmithtt** [comment]
FYI @LPanosTT

**Comment by mywoodstock** [comment]
> @mywoodstock, let us know your thoughts. I think we can probably help make this change if you don't object.

Yes, this is great idea. Please go ahead.

**Comment by nsmithtt** [comment]
Met offline and came up with this pseudo code as a plan:
```c++
Tensor transpose(Tensor tensor);                                                                                                                
                                                                                                                                        
// accepts/returns host tensors                                                                                                         
Tensor prepareWeightsChannelsLastForTTNN(Tensor weightsChannelLast);                                                                
Tensor prepareWeightsChannelsFirstForTTNN(Tensor weightsChannelFirst) {                                                                                                                                               
    return prepareWeightsChannelsLastForTTNN(transpose(weightsChannelFirst));                                                       
}                                                                                                                                                                                                                     
                                                                                                                                                                                                                      
Tensor prepareBiasChannelsLastForTTNN(Tensor biasChannelLast);                                                                      
Tensor prepareBiasChannelsFirstForTTNN(Tensor biasChannelFirst) {                                                                                                                          
    return prepareBiasChannelsLastForTTNN(transpose(biasChannelFirst));                                                               
} 
                                                                                                                           
                                                                                                                                                                                        
std::tuple<Tensor, Tensor, Tensor> conv2dHostWeightsChannelsFirst(Tensor actChannelLast, Tensor hostWeightsChannelFirst, Tensor hostBiasChannelFirst);
                                                                                      
Tensor conv2d(Tensor actChannelLast, Tensor weightsTTNN, Tensor biasTTNN);
```

**Comment by LPanosTT** [comment]
@nsmithtt @mywoodstock The separation of weight and bias preparation into their own ops is complete. The acceptance of different weight formats however is not. I believed we'd need `permute` to work on host for this.

**Comment by nsmithtt** [comment]
@LPanosTT, which format is not supported? Channels last weights?

**Comment by LPanosTT** [comment]
> @LPanosTT, which format is not supported? Channels last weights?

The only format that is supported is OIHW. Which is typically the format of channel-first weights

**Comment by nsmithtt** [comment]
I think ideally we'd like to also support HWIO (i.e. `prepareWeightsChannelsLastForTTNN`), we might need to hand roll this ourselves because TTNN doesn't have a concept of host side ops.

**Comment by nsmithtt** [comment]
This is manually done today, see `to_weight_tile_layout`, we need to add a similar routine (potentially tweak the existing routine) to prepare the weights from a HWIO source layout instead.

**Comment by pavlejosipovic** [comment]
prepare weights is implemented so closing this one.

**Comment by nsmithtt** [comment]
I think this issue was never fully completed.  But at this point things are so entrenched, maybe it's not worth it to fix?

**Comment by LPanosTT** [comment]
@nsmithtt if we get a graph in tt-forge which has a channel last convolution, we will be able to tell from the stablehlo, and insert the proper permutes. This will end up being constevaled anyway.

**Comment by pavlejosipovic** [comment]
@nsmithtt can you than restate the problem at least for my understanding.

**Comment by nsmithtt** [comment]
@pavlejosipovic 

Today the TTNN conv API is a bit inconsistent, typically there are 2 flavors of data layout:

- [PyTorch style (Channels First)](https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.conv2d.html)
  - Activations: `(N, inC, iH, iW)`
  - Weights: `(outC, inC, kH, kW)`
- [Tensorflow style (Channels Last)](https://www.tensorflow.org/api_docs/cc/class/tensorflow/ops/conv2-d)
  - Activations: `(N, iH, iW, iC)`
  - Weights: `(kH, kW, inC, outC)`

This brings us to TTNN API which wants users to input:
- Activations: in Tensorflow (Channels Last) layout
- Weights: in PyTorch (Channels First) layout

It's not a huge deal, but notice that if we went all in on Tensorflow style data layout, the weights (aside from some padding) are basically given to us in the correct layout instead of having to do host routines to convert the weight data from torch layout to tensorflow layout.

**Comment by pavlejosipovic** [comment]
Ok I see. 
Prepared weights are in this channel last fashion, but we would never be able to use existing data as is it would still have to go to prepare for padding/tilize/and reshuffle for blocks sharded convs.
We could consider making unprepared weights in channel last format for consistency.

**Comment by nsmithtt** [comment]
> Ok I see. Prepared weights are in this channel last fashion, but we would never be able to use existing data as is it would still have to go to prepare for padding/tilize/and reshuffle for blocks sharded convs. We could consider making unprepared weights in channel last format for consistency.

My thinking is that if we are given, from the user, channels last style weights then we don't need a specialized conv weight prep path.  I _think_ we can just treat the weights as any other 4 dimensional tensor that needs tile padding on the inner 2 dims.  Perhaps there are 2 special cases when we need them row padded to 4 or 16, but this might also be handled already by just using the small tile sizes.

**Comment by LPanosTT** [comment]
@nsmithtt @pavlejosipovic  the prepare conv weights op requires the shard spec of the would-be activations though, doesn’t it? If so we’d still need this op

**Comment by pavlejosipovic** [comment]
Yes and also layout of the tensor (tiles vs rm) and orientation of the shard. Also we need to tilize the tensor.
We could have a more vanilla weights tensor we would just pay the cost for that in runtime.

**Comment by nsmithtt** [comment]
I see, so it's more involved than I thought.
